### ISSUE TITLE
Remote driver with custom driver not working

### ISSUE THREAD

**Issue by andradah** [zalando/zalenium#403]
If you enable webdriver.remote.url & webdriver.driver then the custom driver, for example, the provided driver is not getting invoked.

To elaborate, I want to execute a test on a remote grid with Zalenium but I want some custom capabilities(extensions) to be set (I need this custom driver for downloading a PDF).

How can I do that?

```
webdriver.remote.url=http://zaleniumhost/wd/hub
webdriver.remote.driver = chrome
webdriver.driver = provided
webdriver.provided.type = mydriver
webdriver.provided.mydriver
=com.serenity.maven.customdrivers.CustomChromeDriver
serenity.driver.capabilities = mydriver
```
If I use the above configuration then CustomChromeDriver class is never called.
#zalenium

**Comment by diemol** [comment]
Hi @andradah,

I don't really understand what you mean, sorry. I am not familiar with this command `webdriver.provided.mydriver`.

Could you please show me how I could use those commands with a standard Selenium Grid? Perhaps after seeing that, I could suggest what should be done with Zalenium.

**Comment by andradah** [comment]
Hi @diemol and thank you for answering, 
I am not sure if it's possible to customize zalenium webdriver in any way.
I tried the customization the way i did with grid, using serenity command: webdriver.provided.mydriver, which invokes a custom webdriver implemented in a java class in my project. 
`package com.tools.customerDrivers;

import java.util.HashMap;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.remote.CapabilityType;
import org.openqa.selenium.remote.DesiredCapabilities;

import net.thucydides.core.webdriver.DriverSource;

public class CustomChromeDriver implements DriverSource {

	@Override
	public WebDriver newDriver() {
		return setCustomChrome();
	}

	@Override
	public boolean takesScreenshots() {
		return true;
	}

	private WebDriver setCustomChrome() {
		String downloadFilepath = System.getProperty("basedir") + "/resources/downloads/";
		HashMap<String, Object> chromePrefs = new HashMap<String, Object>();
		chromePrefs.put("download.default_directory", downloadFilepath);
		chromePrefs.put("plugins.always_open_pdf_externally", true);
		chromePrefs.put("profile.default_content_settings.popups", 0);
		ChromeOptions options = new ChromeOptions();
		options.setExperimentalOption("prefs", chromePrefs);
		options.addArguments("start-maximized");
		options.addArguments("disable-infobars");
		DesiredCapabilities cap = DesiredCapabilities.chrome();
		cap.setCapability(CapabilityType.ACCEPT_SSL_CERTS, true);
		cap.setCapability(ChromeOptions.CAPABILITY, options);

		return new ChromeDriver(cap);
	}
}`
The command i'm using for the test run is this: `mvn verify -DTest=Test_MyTest -DrunPlatform=desktop -DrunEnv=qa-env  -Dwebdriver.driver=provided -Dwebdriver.remote.url=http://myzaleniumhost/wd/hub` and instead of -Dwebdriver.driver=chrome i used provided.
Thank you for your help.

**Comment by diemol** [comment]
Uh, to be honest, I have not used Serenity, so I am not sure how that parameter should be passed. But if it worked with a normal grid, it should work with Zalenium.

Nevertheless, I could take a look and check Serenity briefly during the weekend to see if it is possible.

**Comment by andradah** [comment]
In fact i need to customize the Zalenium browser for downloading documents and then open them for verifications. The downloading part works fine, but i need to download them in a specific folder. That's why i need the customization. The files download in `/home/seluser/Downloads` by default, but apparently this directory is created on the fly in the moment of the download and then test fails when tries to find the directory path `/home/seluser/Downloads`.

**Comment by diemol** [comment]
But you could also specify the download location by using a browser profile.

In the end, what you can do is to mount a local folder to one in the container, and then you can just validate the local folder. 

This example may help you https://github.com/diemol/docker-selenium-issues/blob/master/issue-548/src/test/java/com/elementalselenium/DownloadTest.java

**Comment by diemol** [comment]
Hi @andradah,

I spent some time and create a sample test with Serenity where a file is downloaded, using Zalenium. Check it [here](https://github.com/diemol/docker-selenium-issues/blob/master/zalenium-issue-403/src/test/java/com/example/features/WhenSearchingOnGoogleTest.java)

You would need to start Zalenium like this:

```
    docker run --rm -ti --name zalenium -p 4444:4444 \                                                                                                            
      -v /var/run/docker.sock:/var/run/docker.sock \
      -v /tmp/videos:/home/seluser/videos \
      -v /tmp/downloads:/tmp/node/home/seluser/Downloads \
      --privileged dosel/zalenium start
```

Just note how the mount `-v /tmp/downloads:/tmp/node/home/seluser/Downloads`, which is where the files get downloaded.

I think there is no need to implement a custom driver. 

Let us know how it went.

**Comment by diemol** [comment]
That's strange, I am also using the latest image from `elgalu/selenium` with Chrome version 63.0.3239.132, and the script I linked passes without any issues.

About the docker-compose file, the only strange thing I see is the double mounting:
```
      - /home/volumes:/tmp/node/home/seluser/Downloads
      - /tmp/downloads:/tmp/node/home/seluser/Downloads
``` 

Maybe you can delete the image `elgalu/selenium` completely and pull it again, I have had strange behaviour with some images and after deleting and pulling it again it worked.

**Comment by andradah** [comment]
Hi @diemol, 
I tried to remove and pull again the dosel/zalenium and elgalu/selenium images but the the downloading problem persists. The problem is the Chrome from the container opens that OS window to ask where to download files...
![zaleniumcontainer](https://user-images.githubusercontent.com/20436012/35569806-e061c452-05d5-11e8-8ad3-89a7f966b3c0.png)
Any idea why that happens?

**Comment by andradah** [comment]
Hello again @diemol, 

I just wanted to say that I was able to figure out the issue of downloading. It was not a matter of Zalenium. I run my tests through a Jenkins slave and the chrome download window was opened because somehow the default directory was not allowed to write and read (Chrome does it when the target folder does not have permissions or does not exist). I solved this problem by performing these tests in another Jenkins (with a master only) and zalenium on the same machine.
Thank you for your help, I appreciate :)
